pytorch_lightning/trainer/trainer.py

```diff
@@ -633,14 +635,15 @@ def run_train(self) -> None:
             if not self.interrupted:
                 self.state = TrainerState.INTERRUPTED
                 self.on_keyboard_interrupt()
-        except (RuntimeError, AssertionError):
-            # if an exception is raised, the finally block is executed and can hide the actual exception
-            # that was initially raised if `on_train_end` also raises an exception. we want to avoid that
-            # for assertions and other runtime errors so we aren't misled while debugging
+                self.accelerator.on_train_end()
+                self._running_stage = None
+        except:
+            # give accelerators a chance to finish
+            self.accelerator.on_train_end()
+            # reset bookkeeping
+            self._running_stage = None
             print_exc()
```

Comment:
Probably don't need this print now that we raise again

Reply:
Thanks @carmocca, @awaelchli  and @tchaton for the review, will address the comments and publish.